### PR TITLE
WaitFor: SQL Server

### PR DESCRIPTION
**Pull request by mitchdenny**
## Description

Adds SQL Server support for WaitFor (still needs unit tests and some validation around Azure SQL deployments).

Related: #5645 

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5669)

### CODE REVIEW

File: src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs [hunk 1]
```diff
@@ -28,6 +29,33 @@ public static IResourceBuilder<SqlServerServerResource> AddSqlServer(this IDistr
         var passwordParameter = password?.Resource ?? ParameterResourceBuilderExtensions.CreateDefaultPasswordParameter(builder, $"{name}-password", minLower: 1, minUpper: 1, minNumeric: 1);
 
         var sqlServer = new SqlServerServerResource(name, passwordParameter);
+
+        string? connectionString = null;
+
+        builder.Eventing.Subscribe<ConnectionStringAvailableEvent>(sqlServer, async (@event, ct) =>
+        {
+            connectionString = await sqlServer.GetConnectionStringAsync(ct).ConfigureAwait(false);
```
Comment (radical) [inline src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs:37]: Should this fail if `connectionString` is null, and not publish `ConnectionStringAvailableEvent`?

File: tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs [hunk 2]
```diff
@@ -16,6 +18,101 @@ namespace Aspire.Hosting.SqlServer.Tests;
...
+
+        var rns = app.Services.GetRequiredService<ResourceNotificationService>();
+
+        await rns.WaitForResourceAsync(resource.Resource.Name, KnownResourceStates.Running, cts.Token);
+
+        await rns.WaitForResourceAsync(dependentResource.Resource.Name, KnownResourceStates.Waiting, cts.Token);
+
+        healthCheckTcs.SetResult(HealthCheckResult.Healthy());
+
+        await rns.WaitForResourceAsync(resource.Resource.Name, (re => re.Snapshot.HealthStatus == HealthStatus.Healthy), cts.Token);
```
Comment (radical) [inline tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs:52]: Is there a possible race here - the `ResourceEvent` arrives before we can start watching for the state changes?

File: tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs [hunk 3]
```diff
@@ -16,6 +18,101 @@ namespace Aspire.Hosting.SqlServer.Tests;
...
+
+        await rns.WaitForResourceAsync(dependentResource.Resource.Name, KnownResourceStates.Running, cts.Token);
```
Comment (radical) [inline tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs:54]: And this one - the `state=running` event is received before this line executes?

File: tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs (hunk 2 above)
Comment (davidfowl) [reply tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs:52]: we store the last state and replay it so it *should* work.

**Comment by radical** [comment]
In a separate PR we should add the `WaitForSandbox` playground app to the tests too.

File: tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs [hunk 4]
```diff
@@ -16,6 +18,101 @@ namespace Aspire.Hosting.SqlServer.Tests;
...
+        // Create the database.
+        var connectionString = await resource.Resource.GetConnectionStringAsync(cts.Token);
+        using var connection = new SqlConnection(connectionString);
+        await connection.OpenAsync(cts.Token);
+
+        var command = connection.CreateCommand();
+        command.CommandText = "CREATE DATABASE db;";
+        await command.ExecuteNonQueryAsync(cts.Token);
+
+        await rns.WaitForResourceAsync(db.Resource.Name, (re => re.Snapshot.HealthStatus == HealthStatus.Healthy), cts.Token);
```
Comment (davidfowl) [inline tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs:107]: nit:
```suggestion
        await rns.WaitForResourceAsync(db.Resource.Name, re => re.Snapshot.HealthStatus == HealthStatus.Healthy, cts.Token);
```

File: tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs (hunk 2 above)
Comment (mitchdenny) [reply tests/Aspire.Hosting.SqlServer.Tests/SqlServerFunctionalTests.cs:52]: The `WaitForResourceAsync` event always injects the most recent event when you start watching so it should be fine.

File: src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.SqlServer/SqlServerBuilderExtensions.cs:37]: This is unlikely to occur but I think in the interests of being defensive its not a bad idea to do this.

**Comment by davidfowl** [comment]
@danegsta 

Aspire.Hosting.Dcp.dcpctrl.NetworkReconciler Error: could not create a network	{"NetworkName": {"name":"aspire-network"}, "Reconciliation": 2, "error": "docker command 'CreateNetwork' returned with non-zero exit code 1: command output: Stdout: '' Stderr: 'Error response from daemon: could not find an available, non-overlapping IPv4 address pool among the defaults to assign to the network\n'"}

**Comment by mitchdenny** [comment]
@radical  I reverted the target framework changes. None of the other sames seem to have that.

**Comment by mitchdenny** [comment]
Oh they do on main now. I'm going to merge with main.

**Review by davidfowl — APPROVED** [review]

**Pull request merged**